src/xtgeo/io/tsurf/_tsurf_reader.py

```diff
@@ -0,0 +1,799 @@
+
+    Args:
+        header_name: Surface name from header
+        coord_sys_data: Optional dictionary with coordinate system data
+        vertices: List of vertex coordinates
+        triangles: List of vertex indices
+
+    Returns:
+        TSurfData: Complete TSurf data object
+    """
+
+    # Create header object
+    header = TSurfHeader(name=header_name)
+
+    # Create coordinate system object if data was provided
+    coord_sys = None
+    if coord_sys_data:
+        coord_sys = TSurfCoordSys(
+            name=coord_sys_data["name"],
+            axis_name=coord_sys_data["axis_name"],
+            axis_unit=coord_sys_data["axis_unit"],
+            zpositive=coord_sys_data["zpositive"],
+        )
+
+    # Convert to numpy arrays with appropriate data types
+    vertices_array = np.array(vertices, dtype=np.float64)
+    triangles_array = np.array(triangles, dtype=np.int64)
+
+    return TSurfData(
+        header=header,
+        coord_sys=coord_sys,
+        vertices=vertices_array,
+        triangles=triangles_array,
+    )
+
+
+def _validate_triangulation_data(vertices: np.ndarray, triangles: np.ndarray) -> None:
+    """Basic validation of the triangulation data in the TSurfData object.
+
+    Args:
+        vertices: Array of vertex coordinates
+        triangles: Array of triangle vertex indices
+    """
+
+    # Need at least 3 vertices to form a triangle
+    if vertices.shape[0] < 3:
+        raise ValueError("Less than 3 vertices found in TSurf triangulation data.")
+
+    # Need at least one triangle to form a surface
+    if triangles.size == 0:
+        raise ValueError("No triangles found in TSurf triangulation data.")
+
+    # Check for valid vertex indices in triangles
+    # Must be in the closed range [1, number of vertices]
+    if np.any(triangles < 1):
+        raise ValueError("Triangle vertex indices must be >= 1 in triangulation data.")
+
+    if np.any(triangles > len(vertices)):
+        raise ValueError(
+            "Triangle vertex indices must be <= number of vertices in "
+            "triangulation data."
+        )
+
+
+def _parse_tsurf(
+    stream: Union[TextIOWrapper, StringIO], filepath_errmsg: str
+) -> TSurfData:
+    """
+    Parse a TSurf file from a file stream.
+
+    Args:
+        stream: A stream of the TSurf file.
+        filepath_errmsg: Error context for meaningful error messages
+    """
+
+    # NOTES:
+    # - The TSurf format has many more sections and attributes than
+    #     those currently captured in this parser
+    # - The first line must be the TSurf signature line: 'GOCAD TSurf 1'
+    # - While we are not aware of whether the ordering of the other
+    #     sections is important, the ordering is not strictly enforced
+    #     in this parser
+    # - Only one section of each type is allowed
+    # - The parser raises an error for unknown sections/keywords
+    # - The parser could be extended to handle more sections/keywords
```

Comment:
Put this in the doc string. We use [Google doc strings](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) which have a `Note:` section.